docs: refactor executor in flow (#4548)
jina-ai/jina: docs/fundamentals/executor/executor-in-flow.md

(executor-in-flow)=
# Executors inside a Flow

Executors are a way to group your `DocumentArray` functions and processing logic into a class that can share configuration and state in a modular way.
The main advantage of `Executor`s is using them in a `Flow`, which can serve, scale, and deploy them with ease.

(executor-api)=
## YAML and Python API

An Executor can be loaded from a YAML file or via Python API. 

````{dropdown} executor.py
:open:

> This is a basic Executor, defined in its own file. Notice that there are no {ref}`request bindings <executor-requests>`. These are configured via the Python and YAML syntax.

Reviewer: I would just make it a bit clearer that the 'old' way of defining a request mapping is still valid, and also mention the advantages of doing it this way:

```suggestion
This is a basic Executor, defined in its own file. Notice that there are no {ref}`request bindings <executor-requests>`. If an Executor is used in a Flow, these bindings can be configured via the Python or YAML Flow syntax, so each Flow can use the Executor in its own way. However, the standard way of adding `@requests` in the Executor itself would still work just fine.
```

        super().__init__(**kwargs)
        print(f'parameter_1 = {parameter_1}')
        print(f'parameter_2 = {parameter_2}')

    def my_index(self, docs: DocumentArray, **kwargs):
        print('in my_index, bound to /index')

    def my_search(self, docs: DocumentArray, **kwargs):
        print('in my_search, bound to /search')

    def foo(self, docs: DocumentArray, **kwargs):
        print('in foo, bound to /random')
```

````

This YAML configuration can also be referenced or directly used inside a Flow YAML.
The YAML file has the following format:

````{tab} via Python API

This example shows how to start a Flow with an Executor via the Python API:

```python
with Flow().add(
    uses='MyExecutor',
    uses_with={
      "parameter_1": "foo",
      "parameter_2": "bar"
    },
    uses_metas={
      "name": "MyExecutor",
      "description": "MyExecutor does a thing to the stuff in your Documents",
      "py_modules": ["executor.py"]
    },
    uses_requests={
      "/index": "my_index",
      "/search": "my_search",
      "/random": "foo"
    },
    workspace="some_custom_path"
) as f:
  ...
```

Python API-specific options:

> - `uses` can be a class or string. Defines the Executor to load. You can also run Executors available on [Jina Hub](https://hub.jina.ai) with a {ref}`special syntax <use-hub-executor>`.

Reviewer 2: ```suggestion
- `uses` can be a class or string that defines the Executor to load. You can also run Executors available on [Jina Hub](https://hub.jina.ai) with a {ref}`special syntax <use-hub-executor>`.
```

      "/random": "foo"
    },
    workspace="some_custom_path"
) as f:
  ...
```

- `name`: Name given to the `Executor`. This is dynamically adapted from the `name` in `metas` and depends on some additional arguments like `shard_id`. 
- `replicas`: Number of replicas of the same `Executor` deployed with the `Flow`.
- `shards`: Number of shards of the same `Executor` deployed with the `Flow`.
- `shard_id`: Identifier of the `shard` corresponding to the given `Executor` instance.
- `workspace`: Path to be used by the `Executor`. Note that the actual workspace directory used by the Executor is obtained by appending `'/<executor_name>/<shard_id>/'` to this value.
- `py_modules`: Path to the modules needed to import the `Executor`. This is another way to pass `py-modules` to the `Executor` from the `Flow`
Python API-specific options:

- `uses` can be a class or string. Defines the Executor to load. You can also run Executors available on [Jina Hub](https://hub.jina.ai) with a {ref}`special syntax <use-hub-executor>`.

````{admonition} Note
:class: note
The YAML API will ignore `.runtime_args` during save and load as they are not statically stored
````

### Requests

> By default, an `Executor` object contains `.requests` as an attribute when loaded from the `Flow`. This attribute is a `Dict` describing the mapping between Executor methods and network endpoints: It holds endpoint strings as keys, and pointers to `function`s as values. 

Reviewer 2: Why is `function`s backticked here?

Author: Fixed

>   uses_with:

Reviewer 3: it would be interesting to see here how u can pass some `with` parameters directly in the `MyExecutor`

Author: We wanted to keep it consistent across examples, using `uses_X` in both. In Python you can't use the short form `with`

> - `jtype` is a string. Defines the class name, interchangeable with bang mark `!`;

Reviewer 2: ```suggestion
- `jtype` is a string that defines the class name, interchangeable with bang mark `!`;
```

> Common arguments to both YAML and Python API:

Reviewer 2: ```suggestion
Common arguments for both YAML and Python API:
```

> - `uses_metas` is a key-value map. It defines the meta information of that class. It contains the following fields:

Reviewer 2: Please fix the "Defines" text for each entry, as suggested in prior comments

```yaml
jtype: MyExecutor

> with:

Reviewer 3: we need an example using `with` simply

Author: We wanted to keep it consistent across examples, using uses_X in both. In Python you can't use the short form with

[...]
parameter_1: overriden_parameter_1
parameter_2: overriden_parameter_2
metas: namespace(description='MyExecutor does a thing to the stuff in your Documents', name='Dynamic Name', py_modules='', workspace='overriden_worskpace')
requests: {'/index': <function MyExecutor.foo at 0x7fc3163ddb90>, '/search': <function MyExecutor.bar at 0x7fc3163f0710>}
[...]
```

````

## Internal Executor attributes

When implementing an `Executor`, if your executor overrides `__init__`, it needs to carry `**kwargs` in the signature and call `super().__init__(**kwargs)`
                                 
```python
from jina import Executor

class MyExecutor(Executor):
    def __init__(self, foo: str, bar: int, **kwargs):
        super().__init__(**kwargs)
        self.bar = bar
        self.foo = foo
```

This is important because when an `Executor` is instantiated in the context of a Flow, Jina is adding extra arguments.
Some of these `arguments` can be used when developing the internal logic of the Executor.

These `special` arguments are `workspace`, `requests`, `metas`, `runtime_args`.

> ### `metas`

Author: I'd say the order should be

- workspace
- requests
- metas
- r.. args (these are never configured by a user)

NotImplementedError('no time for it')
```

````

## Graceful shutdown of an Executor

> You might need to execute some logic when your executor's destructor is called.

Reviewer 2: ```suggestion
You might need to execute some logic when your Executor's destructor is called.
```

Process finished with exit code 0
```

````

## Multiple DocumentArrays as input

> You have seen that `Executor` methods can receive 3 types of parameters: `docs`, `parameters` and `docs_matrix`.

Reviewer 2: ```suggestion
You have seen that `Executor` methods can receive three types of parameters: `docs`, `parameters` and `docs_matrix`.
```
